lib/os/thread_posix.cpp

```diff
@@ -44,10 +50,18 @@
 
 namespace pstore {
     namespace threads {
+#    if defined(PSTORE_PTHREAD_SETNAME_NP_1_ARG) || defined(PSTORE_PTHREAD_SETNAME_NP_2_ARGS) ||   \
+        defined(PSTORE_PTHREAD_SETNAME_NP_3_ARGS) || defined(PSTORE_PTHREAD_SET_NAME_NP)
+#        define PTHREAD_SETNAME_NP
+#    endif
 
-#    ifdef __FreeBSD__
+#    if defined(PSTORE_PTHREAD_GETNAME_NP) || defined(PSTORE_PTHREAD_GET_NAME_NP)
+#        define PTHREAD_GETNAME_NP
+#    endif
+
+#    if !defined(PTHREAD_SETNAME_NP) || !defined(PTHREAD_GETNAME_NP)
```

Comment:
Shouldn’t this be && rather that ||?

Reply:
I think it should be ||. As you wrote in your previous comment: 
```I’m concerned about a possible mismatch here. If none of the pthread_setname() APIs that we know about are available, but the getname() API is, then we’ll set the name using a fallback but get it using the API. That wouldn’t be right.```
I agree with your above statement. If either function is missing, we will use the fallback. Therefore, this should use || not &&.

Reply:
> I agree with your above statement.

As do I, but it doesn’t talk about this situation.

>  If either function is missing, we will use the fallback.

If we have a mismatch (where the code knows how to set the name but not how to get it) then this suggests that there’s may be something to fix with the platform support. In this case, I would like the compile to fail with a #error rather than silently fall back to bypassing the use of either function.

Reply:
>
> If we have a mismatch (where the code knows how to set the name but not how to get it) then this suggests that there’s may be something to fix with the platform support. In this case, I would like the compile to fail with a #error rather than silently fall back to bypassing the use of either function.
>

I got your point. However, musl libc library has the mismatch, where the code defines `pthread_setname_np` but not defines `pthread_getname_np`. 
```
-- Performing Test PSTORE_PTHREAD_SETNAME_NP_2_ARGS
-- Performing Test PSTORE_PTHREAD_SETNAME_NP_2_ARGS - Success
-- Performing Test PSTORE_PTHREAD_SETNAME_NP_1_ARG
-- Performing Test PSTORE_PTHREAD_SETNAME_NP_1_ARG - Failed
-- Performing Test PSTORE_PTHREAD_SETNAME_NP_3_ARGS
-- Performing Test PSTORE_PTHREAD_SETNAME_NP_3_ARGS - Failed
-- Performing Test PSTORE_PTHREAD_SET_NAME_NP
-- Performing Test PSTORE_PTHREAD_SET_NAME_NP - Failed
-- Performing Test PSTORE_PTHREAD_GETNAME_NP
-- Performing Test PSTORE_PTHREAD_GETNAME_NP - Failed
-- Performing Test PSTORE_PTHREAD_GET_NAME_NP
-- Performing Test PSTORE_PTHREAD_GET_NAME_NP - Failed
```
There are some discussions about supporting the `pthread_getname_np` in the following links:
https://www.openwall.com/lists/musl/2019/07/17/3 or https://www.openwall.com/lists/musl/2021/05/20/8

The current musl libc still only includes the function `pthread_setname_np`. I think we want to use the fallback code for musl libc to solve the build issue. Should I change code to use fallback code if both functions are not defined or defined __MUSL_ macro? For other mismatch (except musl libc), a compiler error will be given.

Reply:
> The current musl libc still only includes the function pthread_setname_np.

Weird. In that case your existing solution is fine. However, I would leave a FIXME remark explaining why its that way. We can revisit when a version of MUSL supporting getname is released.

Reply:
`FIXME` is added in the commit https://github.com/SNSystems/pstore/pull/93/commits/8d0a265b94f9d10b23f65e77e54aa40edae896be.